rfcs/000-remove-double-relationships.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: Remove double relationships between Neo4j nodes
+- Start Date: 2021-05-25
+- RFC PR: [amundsen-io/rfcs#0000](https://github.com/amundsen-io/rfcs/pull/0000)
+- Amundsen Issue: 
+
+# Remove double relationships between Neo4j nodes
+
+## Summary
+
+Currently all (most?) relationships between Neo4j nodes are modeled with
+two directional edges. This RFC proposes we only use one directional edge
+between the nodes.
+
+## Motivation
+
+The double relationship graph model makes Neo4j queries more complicated
```

Comment:
could you describe this complicate part in more detail? isn't similar to normal rdbms FK and backreference between two models?

Also what is the benefit besides no need to specify two relationships instead of one?

@jinhyukchang done some initial perf testing for duel relationship initially FYI

Reply:
> Also what is the benefit besides no need to specify two relationships instead of one?

I think if there's no performance degradation, and the query expressibility isn't impacted (i.e. it shouldn't be challenging to write the same query as before), then it's purely cruft, no?

Reply:
I am not sure if it is cruft as benchmark is always subject to lots of dimensions. I think the key of this rfc requires to layout the detail of neo4j data migration if we want to go from duel relationship to one.

Reply:
I think cruft is a good way of describing it. It's not really breaking things, it's just sub-optimal, and a bit of a bad design.

There are performance implications, however. Not that you would really notice unless the size of your graph was very, very large. I'm not sure how large the biggest Amundsen deployments are in terms of number of nodes and edges in the Neo4j database, but I'd be surprised if it hit anywhere near where Neo4j performance would actually start suffering due to this issue. Obviously you could also just throw more hardware on it anyway, as implied by @feng-tao.

I'm submitting some changes to the RFC to outline this as soon as I can write it out. Hopefully in an hour or two.

Reply:
Additionally, no, this is not the same as back references in an ORM / relational database. Back references are the most efficient way an ORM framework can effectively refer back to a parent object.

There is no need to create the bi-directional relationships in Neo4j, because you can follow the one relationship without any issues on performance, functionality or impact on maintainability.

The following two Cypher queries return the exact same results.

`
MATCH (n1:TYPE_A)-[r:REL]->(n2:TYPE_B) RETURN n1, r, n2
MATCH (n2:TYPE_B)<-[r:REL]-(n1:TYPE_A) RETURN n1, r, n2
`